app/src/services/gelato.ts

```diff
@@ -0,0 +1,284 @@
+
+    const actionWithdrawLiquidityData = actionWithdrawLiquidityInterface.functions.action.encode([
+      taskData.conditionalTokensAddress,
+      taskData.fpmmAddress,
+      taskData.positionIds,
+      taskData.conditionId,
+      constants.HashZero,
+      taskData.collateralTokenAddress,
+      taskData.receiver,
+    ])
+
+    const action = new GelatoAction({
+      addr: this.addresses.actionWithdrawLiquidity,
+      data: actionWithdrawLiquidityData,
+      operation: Operation.Delegatecall,
+    })
+
+    const task = new Task({
+      conditions: [condition],
+      actions: [action],
+    })
+
+    const expiryDate = 0 // Not expiring
+
+    return gelatoCoreInterface.functions.submitTask.encode([gelatoProvider, task, expiryDate])
+  }
+
+  /**
+   * Cancel pending automated withdrawal task
+   */
+  encodeCancelTask = (taskReceipt: TaskReceipt): string => {
+    const gelatoCoreInterface = new utils.Interface(gelatoCoreAbi)
+    return gelatoCoreInterface.functions.cancelTask.encode([taskReceipt])
+  }
+
+  /**
+   * Whitelist GelatoCore as Module
+   */
+  encodeWhitelistGelatoAsModule = async (): Promise<string> => {
+    const gnosisSafeInterface = new utils.Interface(gnosisSafeAbi)
+    return gnosisSafeInterface.functions.enableModule.encode([this.addresses.gelatoCore])
+  }
+
+  /**
+   * Decode Action Data
+   */
+  decodeSubmitTimeBasedWithdrawalTask = async (hexData: string): Promise<any> => {
+    const data = utils.defaultAbiCoder.decode(
+      ['address', 'address', 'uint256[]', 'bytes32', 'bytes32', 'address', 'address'],
+      utils.hexDataSlice(hexData, 4),
+    )
+    return data
+  }
+
+  /**
+   * Decode Condition Data
+   */
+  decodeTimeConditionData = async (hexData: string): Promise<any> => {
+    const data = utils.defaultAbiCoder.decode(['uint256'], hexData)
+    return data
+  }
+
+  /**
+   * Check if Gnosis Safe whitelisted Gelato Core as a module
+   */
+  isGelatoWhitelistedModule = async (safeAddress: string): Promise<boolean> => {
+    try {
+      const gnosisSafe = new ethers.Contract(safeAddress, gnosisSafeAbi, this.provider)
+      const modules = await gnosisSafe.getModules()
+      let isModule = false
+      modules.forEach((module: string) => {
+        if (utils.getAddress(module) === utils.getAddress(this.addresses.gelatoCore)) isModule = true
+      })
+      return isModule
+    } catch {
+      return false
+    }
+  }
+
+  /**
+   * Returns min collateral amount to schedule auto withdraw
+   */
+  minimumTokenAmount = async (tokenAddress: string, tokenDecimals: number): Promise<number> => {
+    try {
+      const price = await this.findTokenUsdPrice(tokenAddress, tokenDecimals)
+      return GELATO_MIN_USD_THRESH / price
+    } catch (err) {
+      logger.error(`error finding price via uniswap: ${err.message}`)
+      throw new Error(`error finding price via uniswap`)
+    }
+  }
+
+  /**
+   * Returns Uniswap price of collateral to DAI
```

Comment:
Should we maybe use a more stable stablecoin, maybe USDC, to check the USD price? Or maybe there's a reasoning in favour of decentralization here?

Reply:
The assumption was that the difference in exchange rate that DAI and USDC will have will be negligible for the use case, but we can change it to USDC, I'm indifferent about this.

Reply:
Changed to USDC